go/pserver/etcd_lister.go

```diff
@@ -0,0 +1,120 @@
+package pserver
+
+import (
+	"time"
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultEtcdTimeout time.Duration = time.Second * time.Duration(5)
+)
+
+type pserverEtcdLister struct {
+	client    *clientv3.Client
+	timeout   time.Duration
+	endpoints []string
+}
+
+// read ps desired number from etcd.
+func(p *pserverEtcdLister) desired() int {
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		resp, err := p.client.Get(ctx, DefaultPsDesiredPath)
+		cancel()
+		if err != nil {
+			log.Errorf("Get ps dresire number failed! recnnectiong..., %v", err)
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		kvs := resp.Kvs
+		if len(kvs) == 0 {
+			log.Infoln("Waiting for ps desired registered ...")
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		psDesired, err := strconv.Atoi(string(resp.Kvs[0].Value))
+		if err != nil {
+			log.Errorf("psDesired %s invalid %v", psDesired, err)
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		log.Debugf("Get psDesired number: %d\n", psDesired)
+		return psDesired
+	}
+}
+
+func(p *pserverEtcdLister) List() []Server {
+	psDesired := p.desired()
+
+	servers := make([]Server, psDesired)
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		for i := 0; i < psDesired; i++ {
+			psKey := DefaultPsBasePath + strconv.Itoa(i)
+			log.Debugf("checking %s", psKey)
+			resp, err := p.client.Get(ctx, psKey)
+			if err != nil {
+				cancel()
+				log.Infof("Get psKey= %s error, %v", psKey, err)
+				time.Sleep(p.timeout)
+				continue
+			}
+			kvs := resp.Kvs
+			if len(kvs) == 0 {
+				log.Infof("Waiting for ps addr registered ...")
+				time.Sleep(p.timeout)
+				continue
+			}
+
+			psAddr := string(resp.Kvs[0].Value)
+			// TODO(Longfei) check the ps address
+			if  psAddr == "" {
+				cancel()
+				log.Infof("Get psKey = %s,  psAddr is null illegal", psKey, psAddr)
+				time.Sleep(p.timeout)
+				continue
+			}
+			log.Infof("got value (%s) for key: %s", psAddr, psKey)
+			servers[i].Index = i
+			servers[i].Addr = psAddr
+		}
+		cancel()
+		break
+	}
+	return servers
+}
+
+func NewEtcdAddrLister(endpoints string) (Lister, int) {
```

Comment:
It's less abstraction (interface is an abstraction) if the return type is `PserverEtcdLister` (I think it's a good idea to make `pserverEtcdLister` public, perhaps rename to `EtcdLister`, and `desired` public), rather than `Lister`.

The benefit of less abstraction is not very obvious in the beginning, maybe the article below could explain more. I will give a not-so-good-example: At least it's easier to jump to the definition when not returning an interface (editor jumps to the function definition rather than interface definition).

A good general rule in Go would be: [accept interfaces, return structs](https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a)